File: genprime.cs
// Main.cs created with MonoDevelop
// User: steven at 04:10 11/11/2008
//
// To change standard headers go to Edit->Preferences->Coding->Standard Headers
//
using System;
using System.Diagnostics;

namespace GenPrime
{
	class GenPrime
	{
		private static bool isprime(ulong x)
		{
			ulong lim, y;
			if (x < 2)
				return false;
			if (x < 4)
				return true;
			if (x == 5)
				return true;
			if (x % 2 == 0)
				return false;
			if (x % 5 == 0)
				return false;
			if ((x + 1) % 6 != 0)
				if ((x - 1) % 6 != 0)
					return false;
			lim = (ulong)(Math.Sqrt((double)x) + 1.0f);
			for (y = 3; y < lim; y += 2)
			{
				if (x % y == 0)
					return false;
			}
			return true;
		}
		
		private static ulong genprime(ulong max)
		{
			ulong count = 0,
			     current = 1;
			while (count < max)
			{
				if (isprime(current))
					count++;
				current++;
			}
			return current - 1;
		}

		public static void Main(string[] args)
		{
	                ulong start = ulong.Parse(args[0]), stop = ulong.Parse(args[1]) + 1, last;
			Stopwatch sw = new Stopwatch();
	                for ( ulong x = start; x < stop; x += start )
	                {
	                        sw.Start();
				last = genprime(x);
	                        sw.Stop();
				double duration = (double)sw.ElapsedTicks / (double)Stopwatch.Frequency;
				Console.WriteLine( "Found {0,8} primes in {1,10:0.00000} seconds (last was {2,10})",
					x, duration, last);
	                }
		}
	}
}

File: genprime.lua
primes = {}

function isprime(x)
  if x < 2 then
	return false
  end
  if x == 2 then
	return true
  end
  if x % 2 == 0 then
    return false
  end
  if x < 9 then
    return true
  end
  if (x + 1) % 6 ~= 0 then
    if (x - 1) % 6 ~= 0 then
	  return false
	end
  end
  local lim = math.sqrt(x) + 1.0
  local last = 1
  for z=1,#primes,1 do
    last = primes[z]
    if last >= lim then
      return true
    end
    if x % last == 0 then
      return false
    end
  end
  for y=last+2,lim,2 do
    if x % y == 0 then
	  return false
	end
  end
  return true
end

function genprime(mx)
  local count = 0
  local current = 1
  while count < mx do
    if isprime(current) then
      if #primes == 0 or primes[#primes] < current then
        primes[#primes+1] = current
      end
	  count = count + 1
	end
	current = current + 1
  end
  return current - 1
end

function main(startN, endN)
  for x = startN, endN, startN do
    timeS = os.clock()
    lastN = genprime(x)
    timeE = os.clock()
	timeT = timeE - timeS
    io.write(string.format("Found %8d primes in %10.5f seconds (last was %10u)\n", x, timeT, lastN))
  end
end

main(arg[1],arg[2])
